### ISSUE TITLE
Panic error from dovecot 2.2.27 using libressl 2.4.5 (cross-posting at dovecot@dovecot.org)

### ISSUE THREAD

**Issue by hdatma** [libressl-portable/portable#278]
The following results from applying the attached patch to dovecot 2.2.27. 

The system is macOS 10.12.3 with Xcode 8.2.1 and non-Apple clang 3.9.0. 
The configuration is clean (no errors, no warnings). 
The compilation is also clean, apart from a minor issue with var-expand.c. 

This is the result of test_load_v2_key in src/lib-dcrypt/test-crypto.c. 

test_load_v1_public_key .............................................. : ok
Panic: file randgen.c: line 21 (random_fill): assertion failed: (init_refcount > 0)
Error: Raw backtrace: 2   libdcrypt_openssl.so                0x000000010e2cbd04 default_fatal_finish + 36 -> 3   libdcrypt_openssl.so                0x000000010e2cbadd default_fatal_handler + 61 -> 4   libdcrypt_openssl.so                0x000000010e2cc049 i_panic + 169 -> 5   libdcrypt_openssl.so                0x000000010e2f90ec random_fill + 220 -> 6   libdcrypt_openssl.so                0x000000010e2c260d dcrypt_openssl_store_private_key + 1037 -> 7   test-crypto                         0x000000010e23ff54 test_load_v2_key + 580 -> 8   test-crypto                         0x000000010e24190e test_run + 142 -> 9   test-crypto                         0x000000010e23e921 main + 81 -> 10  libdyld.dylib                       0x00007fff8f42f255 start + 1 -> 11  ???                                 0x0000000000000001 0x0 + 1
/bin/sh: line 1: 72426 Abort trap: 6           ./$bin
make[2]: *** [check-test] Error 1
make[1]: *** [check-recursive] Error 1
make: *** [check-recursive] Error 1

The same error also occurs on dovecot 2.2.26.0 with same tools:

[...]
test_load_v1_public_key .............................................. : ok
Panic: file randgen.c: line 21 (random_fill): assertion failed: (init_refcount > 0)
Error: Raw backtrace: 2 libdcrypt_openssl.so 0x0000000105cddd24 default_fatal_finish + 36 -> 3 libdcrypt_openssl.so 0x0000000105cddafd default_fatal_handler + 61 -> 4 libdcrypt_openssl.so 0x0000000105cde069 i_panic + 169 -> 5 libdcrypt_openssl.so 0x0000000105d0b10c random_fill + 220 -> 6 libdcrypt_openssl.so 0x0000000105cd463d dcrypt_openssl_store_private_key + 1037 -> 7 test-crypto 0x0000000105c51f54 test_load_v2_key + 580 -> 8 test-crypto 0x0000000105c5390e test_run + 142 -> 9 test-crypto 0x0000000105c50921 main + 81 -> 10 libdyld.dylib 0x00007fff95dc0255 start + 1 -> 11 ??? 0x0000000000000001 0x0 + 1
/bin/sh: line 1: 24043 Abort trap: 6 ./$bin
make[2]: *** [check-test] Error 1
make[1]: *** [check-recursive] Error 1
make: *** [check-recursive] Error 1

The dovecot developer does not support libressl. 

There are a few alternatives here:
1. someone at libressl fixes dovecot's problem; 
2. openssl stops breaking API compatibilities; 
3. libressl catches up with openssl; 
4. we drop libressl in favour of openssl; 
5. we replace dovecot for a libressl-friendly IMAP server. 

[dcrypt-openssl.c.patch.txt](https://github.com/libressl-portable/portable/files/752441/dcrypt-openssl.c.patch.txt)
[dovecot-openssl-common.c.patch.txt](https://github.com/libressl-portable/portable/files/752442/dovecot-openssl-common.c.patch.txt)

**Comment by yupi2** [comment]
It's panicing because the random number interface isn't initialized. I don't think libressl can affect that. Does the test fail for you when using a clean release with openssl?

**Comment by busterb** [comment]
Correct @yupi2 , the failure isn't in LibreSSL, it's in dovecot's code in 'src/lib/randgen.c' where someone's calling 'random_fill' without calling 'random_init' first. That entire file looks horrifying.

**Comment by busterb** [comment]
To disable Dovecot's builtin random number generator thing, I modified the configure.ac file like so:
```
dnl * find random source
dnl AC_MSG_CHECKING([for /dev/urandom])
dnl if test -c /dev/urandom || test -s /dev/urandom; then
dnl   AC_MSG_RESULT(yes)
dnl   AC_DEFINE(DEV_URANDOM_PATH, "/dev/urandom", [Path to /dev/urandom])
dnl   have_random_source=yes
dnl else
dnl   AC_MSG_RESULT(no)
dnl fi
```
then ran autoconf. This causes Dovecot to then begin actually using LibreSSL's random number generator instead (which happens to never need explicit initialization to work), and the tests pass.

Gonna punt this back to the Dovecot developers. I'd recommend they switch the polarity of their configure tests from using the ad-hoc read-from-/dev/urandom first method to preferring the RAND interfaces first. They're free to also just use the arc4random implementation we provide directly in the OpenBSD tree. This would actually be much more portable these days.

**Comment by busterb** [comment]
to answer your earlier question @yupi2 , Dovecot fails identically with OpenSSL when running 'make check'. I tested the latest OpenSSL 1.0.2k from homebrew:

CPPFLAGS=-I/usr/local//Cellar/openssl/1.0.2k/include LDFLAGS=-L/usr/local//Cellar/openssl/1.0.2k/lib ./configure --with-ssl=openssl

**Comment by sirainen** [comment]
test-dcrypt calls random_init() at the beginning. So does this mean libressl and OpenSSL 1.0.2k added a new conflicting random_init() that overrides Dovecot's?

**Comment by busterb** [comment]
@sirainen not at all

**Comment by busterb** [comment]
The real problem is that Dovecot's build system is linking the object code for src/lib/randgen.c into the test twice, which leads to two copies of random_fill existing, one that is initialized, the other that isn't:
```
random_fill: 1 0x10ecd6f90
random_fill: 0 0x10ed6ae70
```

```
void random_fill(void *buf, size_t size)
{
    size_t pos;
    ssize_t ret;

    fprintf(stderr, "random_fill: %d %p\n", init_refcount, random_fill);
```

Again, this is a Dovecot issue, and I'll stand by the assertion that Dovecot should really just use arc4random_buf where available, and borrow LibreSSL's arc4random compat code where it is not.

**Comment by sirainen** [comment]
Yeah, I figured out that as well after I realized this bug had nothing to do with libressl and everything to do with OSX.

BTW. Dovecot master-branch is not using RAND_*() anymore at all. It uses only /dev/urandom when it wants to have good randomness. For weak randomness it uses arc4random_buf() if it exists - that change was mainly done to avoid warnings in OpenBSD. The weak randomness is used only in places where it doesn't much matter how bad the randomness is (e.g. unit tests).

**Comment by sirainen** [comment]
This is fixed in Dovecot v2.2.28:

Panic: file randgen.c: line 21 (random_fill): assertion failed: (init_refcount > 0)

The rest of the OpenSSL version number checks will be removed in the 2.2.29:

https://github.com/dovecot/core/commit/e47cfc3022ce0cf6e158bb39b83e3a57e7d3130e
https://github.com/dovecot/core/commit/061801f2b390700a18cbefa468a7bab003f4cc8d
